formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt

```diff
@@ -0,0 +1,300 @@
+ */
+@Serializable(with = CborByteStringSerializer::class)
+@ExperimentalSerializationApi
+public class CborByteString(
+    value: ByteArray,
+    vararg tags: ULong
+) : CborPrimitive<ByteArray>(value, tags) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is CborByteString) return false
+        if (!tags.contentEquals(other.tags)) return false
+        return value.contentEquals(other.value)
+    }
+
+    override fun hashCode(): Int {
+        var result = tags.contentHashCode()
+        result = 31 * result + (value.contentHashCode())
+        return result
+    }
+
+    override fun toString(): String {
+        return "CborByteString(" +
+            "tags=${tags.joinToString(prefix = "[", postfix = "]")}, " +
+            "value=h'${value.toHexString()}" +
+            ")"
+    }
+}
+
+/**
+ * Class representing CBOR `null` value
+ */
+@Serializable(with = CborNullSerializer::class)
+@ExperimentalSerializationApi
+public class CborNull(vararg tags: ULong) : CborPrimitive<Unit>(Unit, tags)
+
+/**
+ * Class representing CBOR map, consisting of key-value pairs, where both key and value are arbitrary [CborElement]
+ *
+ * Since this class also implements [Map] interface, you can use
+ * traditional methods like [Map.get] or [Map.getValue] to obtain CBOR elements.
+ */
+@Serializable(with = CborMapSerializer::class)
+@ExperimentalSerializationApi
+public class CborMap(
+    private val content: Map<CborElement, CborElement>,
+    vararg tags: ULong
+) : CborElement(tags), Map<CborElement, CborElement> by content {
+
+    public override fun equals(other: Any?): Boolean =
+        other is CborMap && other.content == content && other.tags.contentEquals(tags)
+
+    public override fun hashCode(): Int = content.hashCode() * 31 + tags.contentHashCode()
+
+    override fun toString(): String {
+        return "CborMap(" +
+            "tags=${tags.joinToString(prefix = "[", postfix = "]")}, " +
+            "content=$content" +
+            ")"
+    }
+
+}
+
+/**
+ * Class representing CBOR array, consisting of CBOR elements.
+ *
+ * Since this class also implements [List] interface, you can use
+ * traditional methods like [List.get] or [List.size] to obtain CBOR elements.
+ */
+@Serializable(with = CborListSerializer::class)
+@ExperimentalSerializationApi
+public class CborList(
+    private val content: List<CborElement>,
+    vararg tags: ULong
+) : CborElement(tags), List<CborElement> by content {
+
+    public override fun equals(other: Any?): Boolean =
+        other is CborList && other.content == content && other.tags.contentEquals(tags)
+
+    public override fun hashCode(): Int = content.hashCode() * 31 + tags.contentHashCode()
+
+    override fun toString(): String {
+        return "CborList(" +
+            "tags=${tags.joinToString(prefix = "[", postfix = "]")}, " +
+            "content=$content" +
+            ")"
+    }
+
+}
```

Comment:
There are few things that are currently missing, but that would otherwise improve DX when it comes to using CborElement API:
- casting functions/properties such as
  - `val CborElement.cborFloat: CborFloat`, `val CborElement.cborList: CborList`
   - `fun CborElement.asCborFloat(): CborFloat`, ...
- builders for aggregates, similar to `JsonObjectBuilder` and `JsonArrayBuilder`

Those could be added later, though.